src/utils/fetchChainLinkPrices.ts

```diff
@@ -0,0 +1,205 @@
+    heartbeat: 27,
+  },
+  {
+    oracleId: 'MATIC',
+    address: '0xab594600376ec9fd91f8e885dadf0ce036862de0',
+    chain: 'polygon',
+    heartbeat: 27,
+  },
+  {
+    oracleId: 'AVAX',
+    address: '0x0a77230d17318075983913bc2145db16c7366156',
+    chain: 'avax',
+    heartbeat: 120,
+  },
+  {
+    oracleId: 'BNB',
+    address: '0x0567f2323251f0aab15c8dfb1967e4e8a7d42aee',
+    chain: 'bsc',
+    heartbeat: 27,
+  },
+  {
+    oracleId: 'LINK',
+    address: '0x2c1d072e956affc0d435cb7ac38ef18d24d9127c',
+    chain: 'ethereum',
+    heartbeat: 3600,
+  },
+  {
+    oracleId: 'USDT',
+    address: '0x3e7d1eab13ad0104d2750b8863b489d65364e32d',
+    chain: 'ethereum',
+    heartbeat: 864000,
+  },
+  {
+    oracleId: 'DAI',
+    address: '0xaed0c38402a5d19df6e4c03f4e2dced6e29c1ee9',
+    chain: 'ethereum',
+    heartbeat: 3600,
+  },
+  {
+    oracleId: 'BTC',
+    address: '0xc907e116054ad103354f2d350fd2514433d57f6f',
+    chain: 'polygon',
+    heartbeat: 27,
+  },
+  {
+    oracleId: 'USDC',
+    address: '0x8fffffd4afb6115b954bd326cbe7b4ba576818f6',
+    chain: 'ethereum',
+    heartbeat: 864000,
+  },
+];
+
+export async function fetchChainLinkPrices(): Promise<Record<string, number>> {
+  const oraclesByChain: Partial<Record<ApiChain, Oracle[]>> = groupBy(oracles, 'chain');
+  const chains = Object.keys(oraclesByChain) as ApiChain[];
+  const pricesPerChain = await Promise.allSettled(
+    Object.entries(oraclesByChain).map(([chain, oracles]) =>
+      fetchPricesForChain(chain as ApiChain, oracles)
+    )
+  );
+
+  const pricesByOracle: Record<string, number[]> = {};
+  for (const i in pricesPerChain) {
+    const chain = chains[i];
+    const result = pricesPerChain[i];
+
+    if (isResultFulfilled(result)) {
+      const prices = result.value;
+      for (const [oracleId, price] of Object.entries(prices)) {
+        if (!pricesByOracle[oracleId]) {
+          pricesByOracle[oracleId] = [];
+        }
+        pricesByOracle[oracleId].push(price);
+      }
+    } else {
+      console.error(`ChainLink: Failed to fetch prices for chain ${chain}`, result.reason);
+    }
+  }
+
+  // Median price for same oracle on multiple chains
+  return mapValues(pricesByOracle, median);
+}
+
+async function fetchPricesForChain(
+  chain: ApiChain,
+  oracles: Oracle[]
+): Promise<Record<string, number>> {
+  const chainId = toChainId(chain);
+  const multicallAddress = MULTICALL_V3[chainId];
+  if (!multicallAddress) {
+    throw new Error(`Multicallv3 address not found for chain ${this.chain}`);
+  }
+  const multicall = new Multicall({
+    web3Instance: web3Factory(chainId),
+    tryAggregate: true,
+    multicallCustomContractAddress: multicallAddress,
+  });
+
+  const { results } = await multicall.call(
+    oracles.map(oracle => ({
+      reference: oracle.oracleId,
+      contractAddress: oracle.address,
+      abi: chainLinkOracleAbi,
+      calls: [
+        {
+          reference: 'roundData',
+          methodName: 'latestRoundData',
```

Comment:
nice!
is this multicall better than the one we usually use?
these reference and parameters looks more verbose

Reply:
This one supports Multicall3 using either web3 or ethers as providers. We are using for treasury and fees too.